### PR TITLE
Plugin: Document the project structure and conventions used in the `lib` folder for PHP code

### PR DESCRIPTION
**Pull request by noisysocks**
## What?
Adds some documentation about our PHP file structure and some best practices when writing PHP.

## Why?
I want to make @tellthemachines happy.

## How?
I wrote it and have not proofread it at all. Please feel free to push changes. It's based off of the discussion in https://github.com/WordPress/gutenberg/issues/33810 and my own gut feelings having merged PHP code from Gutenberg to Core a few times.

## Testing Instructions
Read the file.

### CODE REVIEW

File: lib/README.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+To make it easier for contribtors to know which features need to be merged to
+Core and which features can be deleted, Gutenberg uses the following file
+structure for its PHP code:
+
+- `lib/experimental` - Experimental features that exist only in the plugin. They
+  are not ready to be merged to Core.
+- `lib/stable` - Stable features that exist only in the plugin. They could one
+  day be merged to Core, but not yet.
+- `lib/compat/wordpress-X.Y` - Stable features that are intended to be merged to
```
Comment (ramonjd) [inline lib/README.md:17]: Not suggesting we do it right away, but do  you think it'd be helpful for folks if expanded the compat topic at some stage? For example rationale, preferred naming conventions when extending classes, things to test, and so on. 

There are also plenty of lessons to draw from PRs such as https://github.com/WordPress/gutenberg/pull/38883 and https://github.com/WordPress/gutenberg/pull/38671

Or maybe it's just too fuzzy an area to pin down. 🌪️

**Comment by gziolo** [comment]
@adamziel and @priethor, the first section is important as it covers the recommendations for the plugin project structure. This is something that drives how backporting to WordPress Core new features added in the Gutenberg Plugin.

File: lib/README.md (hunk 1 above)
Comment (noisysocks) [reply lib/README.md:17]: Yes absolutely. This should be a living document, like all our docs. I think we're still figuring a lot of that out, so it's hard to pen down right now. If you have any suggestions on things we can add now then please go ahead!

**Comment by noisysocks** [comment]
Thanks for editing @ramonjd!

File: lib/README.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+For features that may be merged to Core, it's best to use a `wp_` prefix for functions or a `WP_` prefix for classes. 
+
+This applies to both experimental and stable features.
+
+Using the `wp_` prefix facilitates the process of renaming functions and classes from `gutenberg_` to `wp_` if the feature is merged to Core.
+
+Functions that are intended solely for the plugin, e.g., plugin infrastructure, should use the `gutenberg_` prefix.
+
+#### Good
```
Comment (tellthemachines) [inline lib/README.md]: Here we could clarify in what context this example is good.

```suggestion
#### Feature to be merged to Core
```

File: lib/README.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+```php
+function wp_get_navigation( $slug ) { ... }
+```
+
+#### Not so good
```
Comment (tellthemachines) [inline lib/README.md]: Likewise, we could clarify when this should be used:

```suggestion
#### Plugin infrastructure that will never go in Core
```

File: lib/README.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+## Best practices
+
+### Prefer the `wp` prefix
+
+For features that may be merged to Core, it's best to use a `wp_` prefix for functions or a `WP_` prefix for classes. 
+
+This applies to both experimental and stable features.
+
+Using the `wp_` prefix facilitates the process of renaming functions and classes from `gutenberg_` to `wp_` if the feature is merged to Core.
```
Comment (andrewserong) [inline lib/README.md]: Does the phrase `facilitates` here mean "makes it easier to", or "avoids us having to"? It sounds like the suggestion here is to use the `wp_` prefix instead of `gutenberg_` where possible, to reduce the number of functions / classes that need to be renamed? (Just wanted to make sure I understand 😀)

File: lib/README.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+```php
+function gutenberg_get_navigation( $slug ) { ... }
+```
+
+### Group PHP code by _feature_
+
+Developers should organize PHP into files or folders by _feature_, not by _component_.
+
+Also, developers should call `add_action` and `add_filter` immediately after the function being hooked is defined.
```
Comment (andrewserong) [inline lib/README.md]: Totally optional, but I found this sentence a little hard to parse at first. How about:

```suggestion
When defining a function that will be hooked, developers should call `add_action` and `add_filter` immediately after the function declaration.
```

File: lib/README.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+// lib/experimental/init.php
+add_action( 'init', 'wp_register_navigation_cpt' );
+```
+
+### Wrap functions and classes with `! function_exists` and `! class_exists`
+
+Developers should take care to not define functions and classes that are already defined.
+
+When writing new function and classes, it's good practice to use `! function_exists` and `! class_exists`.
```
Comment (andrewserong) [inline lib/README.md:85]: Tiny grammar fix:

```suggestion
When writing new functions and classes, it's good practice to use `! function_exists` and `! class_exists`.
```

Comment (andrewserong) [inline lib/README.md]: Tiny grammar fix:

```suggestion
When writing new functions and classes, it's good practice to use `! function_exists` and `! class_exists`.
```

File: lib/README.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+// lib/experimental/navigation/navigation.php
+
+if ( ! function_exists( 'wp_get_navigation' ) ) {
+	function wp_get_navigation( $slug ) { ... }
+}
+
+// lib/experimental/navigation/class-wp-navigation.php
+
+if ( class_exists( 'WP_Navigation' ) ) {
+	return;
```
Comment (andrewserong) [inline lib/README.md:103]: Nice, TIL you can call return at root level in a PHP file (i.e. not inside a function) to skip execution 👍

**Review by andrewserong — COMMENTED** [review]
Thanks so much for starting this @noisysocks! I've just added a couple questions and a few tiny grammar suggestions, but this looks like a great start for a living document. Nice work!

File: lib/README.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+// lib/experimental/navigation/class-gutenberg-navigation.php
+
+class WP_Navigation { ... }
+```
+
+Furthermore, a quick codebase search will also help you know if your new method is unique.
+
+### Note how your feature should look when merged to Core
+
+Developers should write a brief note about how their feature should be merged to Core, for example, which Core file or function should be patched.
```
Comment (tellthemachines) [inline lib/README.md:125]: This will help so much with backports to Core ⭐

**Review by tellthemachines — APPROVED** [review]
Thanks for writing this up! It's clear and detailed and will make everyone's life easier 🎉

File: lib/README.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
+# Gutenberg PHP
+
+## File structure
+
+Gutenberg adds features to WordPress Core using PHP hooks and filters. Some
+features, once considered stable and useful, are merged into Core during a Core
+release. Some features remain in the plugin forever or are removed.
+
+To make it easier for contribtors to know which features need to be merged to
```
Comment (talldan) [inline lib/README.md]:
```suggestion
To make it easier for contributors to know which features need to be merged to
```

File: lib/README.md (hunk 4 above)
Comment (noisysocks) [reply lib/README.md]: Re-worded this. Your understanding is right!

File: lib/README.md (hunk 5 above)
Comment (noisysocks) [reply lib/README.md]: Sure!

File: lib/README.md (hunk 7 above)
Comment (noisysocks) [reply lib/README.md:103]: PHP is pretty lawless.

**Comment by noisysocks** [comment]
Thanks for the proofreading and suggestions! I'll link to this in the #core-editor chat for awareness. Again, let's iterate on this document. The team is still figuring out a lot of it 😀

File: lib/README.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+function wp_get_navigation( $slug ) { ... }
+```
+
+#### Plugin infrastructure that will never be merged to Core
+
+```php
+function gutenberg_get_navigation( $slug ) { ... }
+```
+
+### Group PHP code by _feature_
```
Comment (youknowriad) [inline lib/README.md:46]: I actually think the `lib/compat` folder should group its code by "core" php file. Meaning if the change needs to happen in `edit-form-blocks.php` in 5.9, I'd put it in `lib/compat/5.9/edit-form-blocks.php` and for new files, try to come up with the best file name that will suite core merge. In practice, this is also very close to "group by feature".

The main goal is to simplify the work for the person back porting.

**Review by youknowriad — COMMENTED** [review]
Really cool addition, thanks Robert.

File: lib/README.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+### Prefer the `wp` prefix
+
+For features that may be merged to Core, it's best to use a `wp_` prefix for functions or a `WP_` prefix for classes. 
+
+This applies to both experimental and stable features.
+
+Using the `wp_` prefix avoids us having to rename functions and classes from `gutenberg_` to `wp_` if the feature is merged to Core.
+
+Functions that are intended solely for the plugin, e.g., plugin infrastructure, should use the `gutenberg_` prefix.
+
```
Comment (ajlende) [inline lib/README.md:33]: Any thoughts on naming conventions for when we're updating a function from core?

For example, this change from https://github.com/WordPress/gutenberg/pull/38681/files?w=1#diff-089523f9c993d53f6f02926d28f7a80394d93740e7d16e47e17244547b2b8f5e. The core action is removed and the plugin action is added to replace it.

My intention with that change was that the `gutenberg_` version of the function in the plugin would replace the `wp_` version when porting to core. If we're reserving that prefix for things that should never go into core, we need another convention.

**Comment by priethor** [comment]
Thanks indeed! These changes helped me better understand the methodology and nuances of the PHP merging process!

File: lib/README.md (hunk 10 above)
Comment (noisysocks) [reply lib/README.md:46]: The downside of that though is that you can't move features in and out of "experimental" status e.g. using `git mv lib/experimental/cool-new-feature.php lib/compat/wordpress-6.0/cool-new-feature.php`. Not sure if we do this in practice, mind you 😀

File: lib/README.md (hunk 11 above)
Comment (noisysocks) [reply lib/README.md:33]: I don't think we really have a good strategy for this kind of stuff. We have a similar nightmare with `WP_Theme_JSON` needing `Gutenberg_Theme_JSON` and `WP_Theme_JSON_5_9` 😅

Using `gutenberg_` seems fine to me, though it will get tricky when the new `gutenberg_` function makes it into WP 6.0 and we want to make more changes in 6.1 We won't be able to tell whether the `wp_` function that exists is the 5.9 version or the 6.0 version.
